google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java

```diff
@@ -568,28 +605,98 @@ public void run() {
+    batch.state = BatchState.SENT;
+    boolean success = rateLimiter.tryMakeRequest(batch.getPendingOperationCount());
     Preconditions.checkState(success, "Batch should be under rate limit to be sent.");
-    try {
-      final ApiFuture<List<BatchWriteResult>> commitFuture = batch.bulkCommit();
-      commitFuture.addListener(
-          new Runnable() {
-            public void run() {
-              try {
-                batch.processResults(commitFuture.get(), null);
-              } catch (Exception e) {
-                batch.processResults(new ArrayList<BatchWriteResult>(), e);
-              }
-              // Remove the batch from BatchQueue after it has been processed.
-              boolean removed = batchQueue.remove(batch);
-              Preconditions.checkState(removed, "The batch should be in the BatchQueue.");
-              sendReadyBatches();
-            }
-          },
-          MoreExecutors.directExecutor());
-    } catch (Exception e) {
-      batch.processResults(new ArrayList<BatchWriteResult>(), e);
+
+    ApiFuture<Void> commitFuture = bulkCommit(batch);
+    commitFuture.addListener(
+        new Runnable() {
+          public void run() {
+            System.out.println("removing batch from batch queue. size: " + batchQueue.size());
+            boolean removed = batchQueue.remove(batch);
+            Preconditions.checkState(
+                removed, "The batch should be in the BatchQueue." + batchQueue.size());
+            sendReadyBatches();
+          }
+        },
+        MoreExecutors.directExecutor());
+  }
+
+  private ApiFuture<List<BatchWriteResult>> invokeBulkCommit(final BulkCommitBatch batch) {
+    return batch.bulkCommit();
+  }
+
+  private ApiFuture<Void> bulkCommit(BulkCommitBatch batch) {
+    return bulkCommit(batch, 0);
+  }
+
+  private ApiFuture<Void> bulkCommit(final BulkCommitBatch batch, final int attempt) {
+    final SettableApiFuture<Void> backoffFuture = SettableApiFuture.create();
+
+    class ProcessBulkCommitCallback implements ApiAsyncFunction<List<BatchWriteResult>, Void> {
+      @Override
+      public ApiFuture<Void> apply(List<BatchWriteResult> results) {
+        batch.processResults(results);
+        if (batch.getPendingOperationCount() > 0) {
+          logger.log(
+              Level.WARNING,
+              String.format(
+                  "Current batch failed at retry #%d. Num failures: %d",
+                  attempt, batch.getPendingOperationCount()));
+
+          if (attempt < MAX_RETRY_ATTEMPTS) {
+            nextAttempt = backoff.createNextAttempt(nextAttempt);
+            BulkCommitBatch newBatch =
+                new BulkCommitBatch(firestore, batch, batch.getPendingDocuments());
+            return bulkCommit(newBatch, attempt + 1);
+          } else {
+            batch.failRemainingOperations(results);
+          }
+        }
+        return ApiFutures.immediateFuture(null);
```

Comment:
This confused me during the review. There is a strange interaction model going on here.

`batch.processResults` changes the internal state, which reduces the number of pending operations (this is not obvious from the callsite). `new BulkCommitBatch` then uses this filtered list of pending operations, but also uses the unfiltered list of writes. 

While this matches what Node does, a lot of the back and forth in Node happens in the same codepath, and here these interactions all cross into and out of UpdateBuilder.

One improvement suggestion:


```suggestion
         batch.processResults(results);
         List<DocumentReference> pendingOps = batch.getRemainingOperations();
        if (!pendingOps.isEmpty()) {
          logger.log(
              Level.WARNING,
              String.format(
                  "Current batch failed at retry #%d. Num failures: %d",
                  attempt,pendingOps.size())));

          if (attempt < MAX_RETRY_ATTEMPTS) {
            nextAttempt = backoff.createNextAttempt(nextAttempt);
            BulkCommitBatch newBatch =
                new BulkCommitBatch(firestore, batch, pendingOps);
            return bulkCommit(newBatch, attempt + 1);
          } else {
            batch.failRemainingOperations(results);
          }
        }
        return ApiFutures.immediateFuture(null);
```